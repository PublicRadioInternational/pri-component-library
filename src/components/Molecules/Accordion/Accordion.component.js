/**
 * @file Accordion.component.js
 * Exports an accordion component.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import styles from './Accordion.css';

import Icon from '../../Atoms/Svg/Icons.component';
import List from '../../Molecules/List/List.component';

const cx = classNames.bind(styles);

/**
 * Component that renders an accordion.
 */
export default class Accordion extends Component {
  static propTypes = {
    accordionList: PropTypes.arrayOf(PropTypes.object),
    color: PropTypes.oneOf(['Blue', 'Orange', 'Yellow', 'Green']),
    accordionTitle: PropTypes.string.isRequired,
    listId: PropTypes.string
  };

  static defaultProps = {
    accordionList: {},
    color: 'Blue',
    listId: null
  };

  state = { revealed: false };

  reveal = e => {
    e.preventDefault();
    this.setState(prevState => ({
      revealed: !prevState.revealed
    }));
  };

  render() {
    const { accordionList, color, accordionTitle, listId } = this.props;
    // Generate a class name based on the color.
    const accordionItemColor = `accordionItem${color}`;
    const accordionTopLinkColor = `accordionTopLink${color}`;
    const accordionContentMenuLinkColor = `accordionContentMenuLink${color}`;
    const accordionTopLinkClasses = cx({
      accordionTopLink: true,
      [accordionTopLinkColor]: color,
      accordionTopLinkOpen: this.state.revealed === true
    });
    const iconClasses = cx({
      iconDown: true,
      iconUp: this.state.revealed === true
    });

    return (
      <div className={`${styles.accordionItem} ${styles[accordionItemColor]}`}>
        <div className={styles.accordionTop} role="tab">
          <h5 className={styles.accordionTopHeading}>
            <a
              className={accordionTopLinkClasses}
              date-toggle={styles.accordionContent}
              date-parent="#accordion"
              href={`#${listId}`}
              aria-expanded="true"
              aria-controls={listId}
              onClick={this.reveal}
            >
              {accordionTitle}
              <Icon
                name="left"
                ariaHidden
                className={iconClasses}
                viewBox="-2 -2 28 28"
              />
            </a>
          </h5>
        </div>
        <List
          id={listId}
          className={styles.accordionContent}
          reveal={this.state.revealed}
          role="tabpanel"
          ariaLabelledby={accordionTitle}
          ulClass={styles.accordionContentMenu}
          classNameOpen={styles.accordionContentOpen}
          liClass={styles.accordionContentMenuItem}
          linkClass={`${styles.accordionContentMenuLink} ${
            styles[accordionContentMenuLinkColor]
          }`}
          listItems={accordionList}
        />
      </div>
    );
  }
}
