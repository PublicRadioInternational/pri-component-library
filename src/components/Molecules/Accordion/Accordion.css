/* import colors... */
@value colors: "../../00_global/colors.css";
@value whiteDark, gray, grayLighter, grayDark, orange, orangeDark, blue, yellow, green from colors;

/* import fonts... */
@value fonts: "../../00_global/fonts.css";
@value heading from fonts;

.accordionItem {
  border-left: 3px solid orange;
  border-top: 1px solid #ddd;
  border-right: 1px solid #ddd;
}

.accordionItem:last-child {
  border-bottom: 1px solid #ddd;
}

.accordionItemBlue {
  border-left-color: blue;
}

.accordionItemYellow {
  border-left-color: yellow;
}

.accordionItemGreen {
  border-left-color: green;
}

.accordionItemOrange {
  border-left-color: orange;
}

.accordionTop {
  cursor: pointer;
}

.accordionTopHeading {
  margin-bottom: 0;
}

.accordionTopLink {
  background: whiteDark;
  color: grayDark;
  display: block;
  padding: 10px 42px 10px 15px;
  font-weight: bold;
  font-size: 1rem;
  text-decoration: none;
  text-transform: uppercase;
  letter-spacing: 0.05rem;
}

.accordionTopLinkOpen {
  background: white;
}

.accordionTopLink:hover {
  background: blue;
  color: white;
}

.iconDown {
  fill: currentcolor;
  float: right;
  margin-right: -30px;
  transition: transform 300ms ease-out;
  transform: rotate(-90deg);
}

.iconUp {
  transform: rotate(90deg);
}

.accordionContent {
  display: none;
}

.accordionContentOpen {
  background-color: white;
  display: block;
}

.accordionContentMenu {
  margin: 0;
  padding: 0;
}

.accordionContentMenuItem {
  border-bottom: 1px dotted grayLighter;
  display: block;
  list-style: none;
  line-height: 1.1;
  padding: 0;
  text-align: left;
}

.accordionContentMenuItem::after {
  content: '';
}

.accordionContentMenuLink {
  color: blue;
  display: block;
  font: 0.875em heading;
  padding: 10px 15px;
}

.accordionContentMenuLink:hover {
  background-color: blue;
  color: white;
}

.accordionTopLinkYellow:hover,
.accordionContentMenuLinkYellow:hover {
  background: yellow;
  color: inherit;
}

.accordionTopLinkOrange:hover,
.accordionContentMenuLinkOrange:hover {
  background: orange;
}

.accordionTopLinkGreen:hover,
.accordionContentMenuLinkGreen:hover {
  background: green;
}
