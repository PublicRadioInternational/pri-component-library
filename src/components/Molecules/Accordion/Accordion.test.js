/**
 * @file Accordion.test.js
 * Contains tests for Accordion.component.js.
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import Accordion from './Accordion.component';

describe('<Accordion />', () => {
  it('Handles accordion open events', () => {
    const wrapper = mount(
      <Accordion
        accordionTitle="Test"
        accordionList={[{ name: 'Google', url: 'http://google.com' }]}
      />
    );
    wrapper.find('a.accordionTopLink').simulate('click');
    expect(wrapper.state().revealed).toEqual(true);
  });

  it('Matches the List snapshot', () => {
    const component = renderer
      .create(
        <Accordion
          accordionTitle="Test"
          accordionList={[
            { name: 'Google', url: 'http://google.com' },
            { name: 'Yahoo', url: 'http://yahoo.com' }
          ]}
        />
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});
