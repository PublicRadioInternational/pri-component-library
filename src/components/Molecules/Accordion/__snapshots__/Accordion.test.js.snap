// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion /> Matches the List snapshot 1`] = `
<div
  className="accordionItem accordionItemBlue"
>
  <div
    className="accordionTop"
    role="tab"
  >
    <h5
      className="accordionTopHeading"
    >
      <a
        aria-controls="collapseNews"
        aria-expanded="true"
        className="accordionTopLink accordionTopLinkBlue"
        date-parent="#accordion"
        date-toggle="accordionContent"
        href="#collapseNews"
        id="newsPrograms"
        onClick={[Function]}
      >
        Test
        <svg
          aria-hidden={true}
          aria-label={null}
          aria-labelledby={null}
          className="svg iconDown"
          fill="currentcolor"
          height={28}
          version={1.1}
          viewBox="-2 -2 28 28"
          width={28}
        >
          <path
            d="M14.414 5.586c-0.78-0.781-2.048-0.781-2.828 0l-6.415 6.414 6.415 6.414c0.39 0.391 0.902 0.586 1.414 0.586s1.024-0.195 1.414-0.586c0.781-0.781 0.781-2.047 0-2.828l-3.585-3.586 3.585-3.586c0.781-0.781 0.781-2.047 0-2.828z"
          />
        </svg>
      </a>
    </h5>
  </div>
  <div
    aria-labelledby="newsPrograms"
    className="accordionContent"
    role="tabpanel"
  >
    <ul
      className="list accordionContentMenu"
    >
      <li
        className="listItem accordionContentMenuItem"
      >
        <a
          className="listLink accordionContentMenuLink accordionContentMenuLinkBlue "
          href="http://google.com"
        >
          Google
        </a>
      </li>
      <li
        className="listItem accordionContentMenuItem"
      >
        <a
          className="listLink accordionContentMenuLink accordionContentMenuLinkBlue "
          href="http://yahoo.com"
        >
          Yahoo
        </a>
      </li>
    </ul>
  </div>
</div>
`;
